sdl_android/src/main/java/com/smartdevicelink/api/screen/TextAndGraphicManager.java

```diff
@@ -0,0 +1,893 @@
+		textField4Type = null;
+		mediaTrackTextField = null;
+		textAlignment = null;
+		primaryGraphic = null;
+		secondaryGraphic = null;
+		blankArtwork = null;
+		displayCapabilities = null;
+		inProgressUpdate = null;
+		queuedImageUpdate = null;
+		currentScreenData = null;
+		queuedUpdateListener = null;
+		pendingHMIListener = null;
+		inProgressListener = null;
+		hasQueuedUpdate = false;
+		isDirty = false;
+		pendingHMIFull = false;
+
+		// remove listeners
+		internalInterface.removeOnRPCNotificationListener(FunctionID.ON_HMI_STATUS, hmiListener);
+		internalInterface.removeOnSystemCapabilityListener(SystemCapabilityType.DISPLAY, onDisplayCapabilitiesListener);
+
+		// transition state
+		transitionToState(SHUTDOWN);
+	}
+
+	private void addListeners() {
+		// add listener
+		hmiListener = new OnRPCNotificationListener() {
+			@Override
+			public void onNotified(RPCNotification notification) {
+				currentHMILevel = ((OnHMIStatus)notification).getHmiLevel();
+				if (currentHMILevel == HMILevel.HMI_FULL){
+					if (pendingHMIFull){
+						Log.v(TAG, "Acquired HMI_FULL with pending update. Sending now");
+						pendingHMIFull = false;
+						sdl_update(pendingHMIListener);
+						pendingHMIListener = null;
+					}
+				}
+			}
+		};
+		internalInterface.addOnRPCNotificationListener(FunctionID.ON_HMI_STATUS, hmiListener);
+
+		// Add OnDisplayCapabilitiesListener to keep displayCapabilities updated
+		onDisplayCapabilitiesListener = new OnSystemCapabilityListener() {
+			@Override
+			public void onCapabilityRetrieved(Object capability) {
+				displayCapabilities = (DisplayCapabilities)capability;
+			}
+
+			@Override
+			public void onError(String info) {
+				Log.w(TAG, "DISPLAY Capability cannot be retrieved:");
+				displayCapabilities = null;
+			}
+		};
+		this.internalInterface.addOnSystemCapabilityListener(SystemCapabilityType.DISPLAY, onDisplayCapabilitiesListener);
+	}
+
+	// Upload / Send
+
+	protected void update(CompletionListener listener) {
+
+		// check if is batch update
+		if (batchingUpdates) {
+			return;
+		}
+
+		if (isDirty){
+			isDirty = false;
+			sdl_update(listener);
+		}
+	}
+	
+	private void sdl_update(CompletionListener listener){
+
+		// make sure hmi is not none
+		if (currentHMILevel == null || currentHMILevel == HMILevel.HMI_NONE){
+			Log.v(TAG, "Trying to send show on HMI_NONE, waiting for full");
+			pendingHMIFull = true;
+			if (listener != null){
+				pendingHMIListener = listener;
+			}
+			return;
+		}
+
+		Log.v(TAG, "Updating Text and Graphics");
+		if (inProgressUpdate != null){
+
+			Log.v(TAG, "In progress update exists, queueing update");
+			if (queuedUpdateListener != null){
+
+				Log.v(TAG, "Queued update already exists, superseding previous queued update");
+				queuedUpdateListener.onComplete(false);
+				queuedUpdateListener = null;
+			}
+
+			if (listener != null){
+				queuedUpdateListener = listener;
+			}else{
+				hasQueuedUpdate = true;
+			}
+			return;
+		}
+
```

Comment:
This is the point where the synchronization block should start until the end of the method

Reply:
I added the `synchronized` modifier to the entire method as described in a previous comment